### PR TITLE
PencilBrush SHIFT/straight line functionality

### PR DESCRIPTION
**Pull request by ShaMan123**
This PR adds the ability to draw a straight line when pressing the `shift` key or when setting `drawStraightLine` to `true`.

```js
// Override `shift` functionality
brush._detach();

// Restore `shift` functionality
brush._attachKeyboardListeners();

// direct setter
brush.drawStraightLine = true;
```

### Logic
When `drawStraightLine = true` the brush simply pops the last point and adds the new one

### [Live Example](https://codesandbox.io/s/blue-shape-evc6t?file=/src/ShiftBrush.js:10606-11800)

### CODE REVIEW

File: src/brushes/pencil_brush.class.js [hunk 1]
```diff
@@ -21,6 +35,37 @@
...
+          _this.drawStraightLine = true;
+        }
+      };
+      var keyboardUp = function () {
+        _this.drawStraightLine = false;
+      };
+
+      window.addEventListener('keydown', keyboardDown);
+      window.addEventListener('keyup', keyboardUp);
+      this._detach = function disposer() {
```
Comment (ShaMan123) [inline src/brushes/pencil_brush.class.js:60]: Should we call `_detach` somewhere once the brush is dead? and if so from where?

File: src/brushes/pencil_brush.class.js [hunk 2]
```diff
@@ -246,7 +296,7 @@
       var zoom = this.canvas.getZoom(), adjustedDistance = Math.pow(distance / zoom, 2),
           i, l = points.length - 1, lastPoint = points[0], newPoints = [lastPoint],
           cDistance;
-      for (i = 1; i < l; i++) {
+      for (i = 1; i <= l; i++) {
```
Comment (ShaMan123) [inline src/brushes/pencil_brush.class.js:299]: Needs to be validated against #6966

File: src/brushes/pencil_brush.class.js (hunk 1 above)
Comment (asturur) [reply src/brushes/pencil_brush.class.js:60]: The event for SHIFT can be grabbed from the normal mouse event.
SHIFT, CTRL, ALT are modifier keys and can checked on the mousedown or mousemove.

Comment (ShaMan123) [reply src/brushes/pencil_brush.class.js:60]: So you say no to attaching keyboard events.
Should I remove `_attachKeyboardListeners` entirely?

Comment (ShaMan123) [reply src/brushes/pencil_brush.class.js:63]: ed43f43

Comment (ShaMan123) [reply src/brushes/pencil_brush.class.js:63]: I have changed events to listen to canvas

**Comment by ShaMan123** [comment]
this should be merged

**Comment by ShaMan123** [comment]
@asturur please re open and merge

File: src/brushes/pencil_brush.class.js [hunk 3]
```diff
@@ -21,6 +35,41 @@
...
+      this.canvas.on('mouse:down', setShift);
+      this.canvas.on('mouse:move', setShift);
+      this.canvas.on('mouse:up', revokeShift);
+      this._detach = function disposer() {
+        this.canvas.off('mouse:down', setShift);
+        this.canvas.off('mouse:move', setShift);
+        this.canvas.off('mouse:up', revokeShift);
+      };
+      return this._detach;
+    },
```
Comment (asturur) [inline src/brushes/pencil_brush.class.js:69]: ok sorry i took my time to review this, but there was a whole thing i wanted to try.
In my opinion all this new function is not needed. You do not need to attach keyboard events at all to grab the shift key.

Comment (asturur) [reply src/brushes/pencil_brush.class.js:69]: https://github.com/fabricjs/fabric.js/pull/7034/files#diff-62b861bab6725568506ad50661f0be7c7d3d79b8ba69229f664226fb54d9269bR93

Can we add here for example:
```
this.drawStraightLine = options.e[this.straightLineKey];
```

and same on mouseMove / mouseUp.

Keeping `this.straightLineKey` default to 'shiftKey' will allow people to disable the functionality setting it to empty space or a key that does not exist.

Comment (ShaMan123) [reply src/brushes/pencil_brush.class.js:40]: 54f4272
Works!

File: src/brushes/pencil_brush.class.js (hunk 2 above)
Comment (ShaMan123) [reply src/brushes/pencil_brush.class.js:277]: It is why there's a merge conflict

File: src/brushes/pencil_brush.class.js [hunk 4]
```diff
@@ -88,6 +110,7 @@
       if (!this.canvas._isMainEvent(options.e)) {
         return true;
       }
+      this.drawStraightLine = options.e[this.straightLineKey];
```
Comment (asturur) [inline src/brushes/pencil_brush.class.js:113]: wasn't this false by definition?

Comment (ShaMan123) [reply src/brushes/pencil_brush.class.js:113]: can you rephrase?

File: src/brushes/pencil_brush.class.js (hunk 2 above)
Comment (asturur) [reply src/brushes/pencil_brush.class.js:277]: so master is at < l -1 to fix a tail issue with decimate.
Why did you add the equal in?
If you explain we can solve the conflict.
But to me is not clear why this has changed.

Comment (ShaMan123) [reply src/brushes/pencil_brush.class.js:277]: The logic of straight line pops the last point and push the last pointer.
When mouse up fires and `drawStraightLine=true` it decimated the end and removes the last straight line part.
I don't know know/understand what decimate does so I don't want to propose anything but this can be mitigated by pushing another point on mouse up if `drawStraightLine=true` (I don't like it though)

Comment (ShaMan123) [reply src/brushes/pencil_brush.class.js:277]: You can see/test
https://codesandbox.io/s/blue-shape-evc6t?file=/src/ShiftBrush.js:9550-9576

File: src/brushes/pencil_brush.class.js (hunk 4 above)
Comment (ShaMan123) [reply src/brushes/pencil_brush.class.js:113]: Do mean this?
https://github.com/fabricjs/fabric.js/pull/7034/files/54f4272aca08187618ade6ed8539cf763678378d#diff-62b861bab6725568506ad50661f0be7c7d3d79b8ba69229f664226fb54d9269bR23

Comment (asturur) [reply src/brushes/pencil_brush.class.js:113]: Weren't you setting this as false on mouse up, always? is not important for me, i just want to understand

Comment (ShaMan123) [reply src/brushes/pencil_brush.class.js]: It is good practice I guess. 
16fd4e5

**Comment by asturur** [comment]
Sorry this PR was good after my conflict resolve, eventually waiting for tests to pass i forgot about it.

**Review by asturur — APPROVED** [review]
